### PR TITLE
[DI] fix combinatorial explosion when analyzing the service graph

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #29336
| License       | MIT
| Doc PR        | -

and a few minor things found meanwhile.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -155,17 +155,18 @@ public function dump(array $options = array())
...
+        foreach ($this->container->getCompiler()->getServiceReferenceGraph()->getNodes() as $id => $node) {
+            if (!$node->getValue() instanceof Definition) {
+                continue;
+            }
+            if (!isset($checkedNodes[$id])) {
+                $this->analyzeCircularReferences($id, $node->getOutEdges(), $checkedNodes);
             }
         }
         $this->container->getCompiler()->getServiceReferenceGraph()->clear();
+        $checkedNodes = array();
```
Comment (alexander-schranz) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: looks unused?  should it be `unset($checkedNodes);`?

**Review by alexander-schranz — COMMENTED** [review]
@nicolas-grekas thank you! The performance is now again near the 4.1.7 version:

Current benchmark:

v4.1.7: **8,85s** user 1,02s system 98% cpu 9,987 total
v4.1.8: **68,62s** user 1,42s system 98% cpu 1:05,01 total
v4.1.8 with this patch: **11,52s** user 1,29s system 94% cpu 13,576 total

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: could be but that would provide anything more so not really needed

**Comment by nicolas-grekas** [comment]
@alexander-schranz thanks for the confirmation.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 1 above)
Comment (alexander-schranz) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: unset would avoid IDE or code style checker marking it as unused variable so nobody would remove it accidentally ;).

But I'm happy that the container now again only take some seconds to build 😄

**Pull request merged**